src/Compilers/Core/Portable/RealParser.cs

```diff
@@ -0,0 +1,783 @@
+                Debug.Assert((shiftedExponent & ~(((1UL << this.ExponentBits) - 1) << DenormalMantissaBits)) == 0); // exponent fits in its place
+                result = shiftedExponent | mantissa;
+                return SLD_STATUS.SLD_OK;
+            }
+        }
+
+        /// <summary>
+        /// Properties of a C# float.
+        /// </summary>
+        private class FloatFloatingPointType : FloatingPointType
+        {
+            public static FloatFloatingPointType Instance = new FloatFloatingPointType();
+            private FloatFloatingPointType() { }
+            public override ushort DenormalMantissaBits => 23;
+            public override ushort ExponentBits => 8;
+            public override int MaxBinaryExponent => 127;
+            public override int ExponentBias => 127;
+            public override ulong Zero => FloatToInt32Bits(0.0f);
+            public override ulong Infinity => FloatToInt32Bits(1.0f / 0.0f);
+        }
+
+        /// <summary>
+        /// Properties of a C# double.
+        /// </summary>
+        private class DoubleFloatingPointType : FloatingPointType
+        {
+            public static DoubleFloatingPointType Instance = new DoubleFloatingPointType();
+            private DoubleFloatingPointType() { }
+            public override ushort DenormalMantissaBits => 52;
+            public override ushort ExponentBits => 11;
+            public override int MaxBinaryExponent => 1023;
+            public override int ExponentBias => 1023;
+            public override ulong Zero => (ulong)BitConverter.DoubleToInt64Bits(0.0d);
+            public override ulong Infinity => (ulong)BitConverter.DoubleToInt64Bits(1.0d / 0.0d);
+        }
+
+        /// <summary>
+        /// This type is used to hold a partially-parsed string representation of a  
+        /// floating point number.  The number is stored in the following form:  
+        ///  <pre>
+        ///     0.Mantissa * 10^Exponent
+        ///  </pre>
+        /// The Mantissa buffer stores the mantissa digits as characters in a string.  
+        /// The MantissaCount gives the number of digits present in the Mantissa buffer.
+        /// There shall be neither leading nor trailing zero digits in the Mantissa.
+        /// </summary>
+        [DebuggerDisplay("0.{Mantissa}e{Exponent}")]
+        private struct DecimalFloatingPointString
+        {
+            public int Exponent;
+            public string Mantissa;
+            public uint MantissaCount => (uint)Mantissa.Length;
+
+            /// <summary>
+            /// Create a DecimalFloatingPointString from a string representing a floating-point literal.
+            /// </summary>
+            /// <param name="source">The text of the floating-point literal</param>
+            public static DecimalFloatingPointString FromSource(string source)
+            {
+                var mantissaBuilder = new StringBuilder();
```

Comment:
How about some 'early outs' for common cases like "0.0" or "1.0" (which probably occur a lot in generated code, say for WPF)